code/entities/machine/soda-fountain/interactable/Dispenser.cs

```diff
@@ -0,0 +1,93 @@
+﻿using Sandbox;
+using Sandbox.util;
+using Sandbox.Utility;
+
+namespace Cinema;
+
+public partial class Dispenser : BaseInteractable
+{
+    static private float DispenseTime = 3.0f;
+    
+    private RealTimeUntil TimeUntilFinishedDispensing = 0.0f;
+    public bool IsDispensing { get; set; }
+
+    public string AnimationName { get; set; }
+    public SodaFountain.SodaType SodaType { get; set; }
+    private Particles SodaParticles { get; set; }
+    private FillableCup Cup { get; set; }
+
+    public Dispenser() // For the compiler...
+    {
+        
+    }
+
+    /// <summary>
+    /// 
+    /// </summary>
+    /// <param name="cup"></param>
+    public void SetCup(FillableCup cup)
+    {
+        Cup = cup;
+    }
+
+    /// <summary>
+    /// 
+    /// </summary>
+    /// <param name="player"></param>
+    public override void Trigger(Player player) {
+        // If an assembled cup is underneath this dispenser,
+        // don't dispense until it is picked up
+        if (IsDispensing || (Cup.IsValid() && Cup.IsAssembled))
+            return;
+
+        string particlePath;
+
+        if (Cup.IsValid())
+        {
+            particlePath = SodaFountain.GetSodaFillParticlePath(SodaType);
+        }
+        else
+        {
+            particlePath = SodaFountain.GetSodaDispenseParticlePath(SodaType);
+        }
+
+        SodaParticles = Particles.Create(particlePath, Parent);
+        SodaParticles.SetEntityAttachment(0, Parent, Attachment);
+
+        FillCup();
+    }
+
+    public async void FillCup()
+    {
+        IsDispensing = true;
```

Comment:
Race condition safety.

Move `IsDispening = true` into `Trigger()`

Since `FillCup` is async and you are not awaiting it, you shouldn't assume that it will execute immediately.

Ideally you should move all the code in this function before the `await` into the caller (`Trigger`).

Reply:
Moved the code before await into the trigger function